src/mbgl/style/expression/distance.cpp

```diff
@@ -137,25 +190,87 @@ double pointsToPointsDistance(const mapbox::geometry::multi_point<double>& point
+double pointToPolygonDistance(const mapbox::geometry::point<double>& point,
+                              const mapbox::geometry::polygon<double>& polygon,
+                              mapbox::cheap_ruler::CheapRuler& ruler) {
+    if (pointWithinPolygon(point, polygon, true /*trueOnBoundary*/)) return 0.0;
+    double dist = InfiniteDistance;
+    for (const auto& ring : polygon) {
+        const auto nearestPoint = std::get<0>(ruler.pointOnLine(ring, point));
+        dist = std::min(dist, ruler.distance(point, nearestPoint));
+        if (dist == 0.0) return dist;
+    }
+    return dist;
+}
+
+double lineToPolygonDistance(const mapbox::geometry::line_string<double>& line,
+                             const IndexRange& range,
+                             const mapbox::geometry::polygon<double>& polygon,
+                             mapbox::cheap_ruler::CheapRuler& ruler) {
+    if (!isRangeSafe(range, line.size())) {
+        mbgl::Log::Error(mbgl::Event::General, "index is out of range");
+        return InvalidDistance;
+    }
+    for (std::size_t i = range.first; i <= range.second; ++i) {
+        if (pointWithinPolygon(line[i], polygon, true /*trueOnBoundary*/)) return 0.0;
+    }
+
+    double dist = InfiniteDistance;
+    for (std::size_t i = range.first; i < range.second; ++i) {
+        const auto& p1 = line[i];
+        const auto& p2 = line[i + 1];
+        for (const auto& ring : polygon) {
+            for (std::size_t j = 0; j < ring.size() - 2; ++j) {
+                const auto& q1 = ring[j];
+                const auto& q2 = ring[j + 1];
+                if (segmentIntersectSegment(p1, p2, q1, q2)) return 0.0;
+                dist = std::min(dist, segmentToSegmentDistance(p1, p2, q1, q2, ruler));
+            }
         }
-        auto size1 = size / 2;
-        IndexRange range1(range.first, range.first + size1);
-        IndexRange range2(range.first + size1, range.second);
-        return std::make_pair(std::move(range1), std::move(range2));
-    } else {
-        if (size == 1) {
-            return std::make_pair(range, nullopt);
+    }
+    return dist;
+}
+
+double polygonToPolygonDistance(const mapbox::geometry::polygon<double>& polygon1,
+                                const mapbox::geometry::polygon<double>& polygon2,
+                                mapbox::cheap_ruler::CheapRuler& ruler,
+                                double currentMiniDist = InfiniteDistance) {
+    const auto bbox1 = getBBox(polygon1);
+    const auto bbox2 = getBBox(polygon2);
+    if (currentMiniDist != InfiniteDistance && bboxToBBoxDistance(bbox1, bbox2, ruler) >= currentMiniDist) {
+        return currentMiniDist;
+    }
+    const auto polygonIntersect = [](const mapbox::geometry::polygon<double>& poly1,
+                                     const mapbox::geometry::polygon<double>& poly2) {
+        for (const auto& ring : poly1) {
+            for (std::size_t i = 0; i <= ring.size() - 2; ++i) {
```

Comment:
Is it possible to have a ring with less than 2 points? `ring.size() - 2` would cause integer underflow in those scenarios?

Reply:
validity check added